### PR TITLE
Make shared library produced by cmake and configure identical

### LINKED ISSUE #638: Hard to verify that cmake and configure do the same thing
The two build systems do not produce exactly the same output, and it's hard to know if that is benign or a real problem.  But it turns out to be a real problem; see e.g. https://github.com/zlib-ng/zlib-ng/issues/553

### PR DESCRIPTION
**Pull request by dankegel**
zlib shared libraries really ought to be identical regardless of whether you build them with cmake or configure.
This pull request verifies that, and makes it so, on four platforms, at least for plain old and --zlib-compat builds.

Here's what it took (beyond the static library fix landed recently):
- canonicalize CHOST in configure
- rearrange source file order in cmake and configure to match (at least for i686, amd64, aarch64, ppc, and ppc64le)
- don't link shared libraries explicitly with -lc (shouldn't be needed; caused differences between configure and cmake)
- on mac, do like configure, and don't use @rpath
- update pkgcheck.sh to verify that cmake and configure create identical shared libraries

Along the way, it turned out that the configure sections for 32 and 64 bit x86 were
essentially identical, so this pull request merges them to reduce chance of
accidental divergence later.

Fixes #638 and #666 

Known issues left for future pull requests:
- There may still be differences between cmake and configure on architectures not yet tested by github/workflows/pkgcheck.yaml
- --native and -DWITH_NATIVE_INSTRUCTIONS don't do quite the same thing (see #697)

### CODE REVIEW

**Comment by dankegel** [comment]
Rebased.

**Comment by dankegel** [comment]
This is ready to go, I think.

**Comment by nmoinvaz** [comment]
I think the problem with configure script for PPC detection when cross compile should be fixed.

And configure and cmake should be set to -O2. 

That should fix some workarounds that were needed I believe, also you won’t need to check CMakeCache.txt.

File: configure
Comment (Dead2) [inline configure]: Duplicates -DX86_CPUID from the above block?

File: test/pkgcheck.sh
Comment (Dead2) [inline test/pkgcheck.sh]: Perhaps fix this powerpc-detection in configure instead?

File: test/pkgcheck.sh
Comment (mscastanho) [inline test/pkgcheck.sh]: The inner part of that for loop could be simplified to just `mv $a ${a%%.*}.o`. But to be honest I'm not sure if that's portable to other shells too. If not, just disconsider.

File: .github/workflows/pkgcheck.yml
Comment (mscastanho) [inline .github/workflows/pkgcheck.yml]: I believe we should switch this to PPC64LE. This is the ABI followed by recent POWER processors and I believe not many recent distros still support ppc or ppc64, so if we need to choose one, ppc64le is the way to go since it will cover most of the current ppc user base.

File: test/pkgcheck.sh
Comment (mscastanho) [inline test/pkgcheck.sh]: Some as I commented above. Probably better to use ppc64le.

File: test/pkgcheck.sh
Comment (mscastanho) [reply test/pkgcheck.sh]: I believe the issue is on `configure:26`:
https://github.com/zlib-ng/zlib-ng/blob/9abe04b926ef688efa3d4f464209246a22afb01d/configure#L25-L28
When `CHOST` is something like `powerpc-linux-gnu`, it will match the second pattern, which will select the last string after the `-`. But this issue happens with any `CHOST` value in the form `first-second-third`, such as `arm-linux-gnueabihf` and `aarch64-linux-gnu`. All these match the second pattern.

That change may deserve a separate PR though. For now I believe setting `--uname=linux` will suffice.

File: configure
Comment (mscastanho) [inline configure:74]: Looks good. If somebody actually needs `-lc` they can just set it before calling configure.

**Review by mscastanho — COMMENTED** [review]
Some comments here and there but overall the PR looks good. My main suggestion is to use ppc64le instead of ppc in the CI tests.

**Comment by mscastanho** [comment]
Also an additional comment in the `Usage` string of `test/pkgcheck.sh` saying that the example commands are for Ubuntu/Debian-like distros would be nice. Since the instructions use `apt` and the names of the dependency packages may change on other distros. Something simple like `For example, on Ubuntu:`.

File: test/pkgcheck.sh
Comment (mscastanho) [reply test/pkgcheck.sh]: Just opened a new issue about this so it doesn't get forgotten: https://github.com/zlib-ng/zlib-ng/issues/666

File: configure
Comment (dankegel) [reply configure]: Thanks.

File: .github/workflows/pkgcheck.yml
Comment (dankegel) [reply .github/workflows/pkgcheck.yml]: Happy to add ppc64le, but isn't a lot of embedded powerpc going to be 32 bit for the foreseeable future?  So maybe keep the 32 bit one... I have a Powermac G4, and I'm biased :-)

File: test/pkgcheck.sh
Comment (dankegel) [reply test/pkgcheck.sh]: I added a commit to address this.

File: test/pkgcheck.sh
Comment (dankegel) [reply test/pkgcheck.sh]: That syntax looks strange to me :-)  But yes, that seems portable and would be shorter.  If I get a vote or two 
for the %% syntax, I'll switch to it.

File: configure
Comment (dankegel) [reply configure:74]: Agreed.

File: test/pkgcheck.sh
Comment (dankegel) [reply test/pkgcheck.sh]: Added.

**Comment by dankegel** [comment]
Resolved except maybe for the %% and the "on ubuntu" wording in test/pkgcheck.sh comments.
Tomorrow morning I'll push an update for those and anything else that comes up tonight.

**Comment by nmoinvaz** [comment]
@dankegel have you encountered the issue with `-mfpu=neon` being set on all files when Neon is present in configure, but in CMake we only set it per file that uses neon?

**Comment by dankegel** [comment]
Yes.  That's why I wimped out and only test aarch64, not 32 bit arm...

**Comment by nmoinvaz** [comment]
See #669 and #668. Maybe you can cherry pick those and see if they solve the problems?

File: .github/workflows/pkgcheck.yml
Comment (mscastanho) [reply .github/workflows/pkgcheck.yml]: Agreed. The more the merrier =)

File: .github/workflows/pkgcheck.yml
Comment (mscastanho) [inline .github/workflows/pkgcheck.yml]: @dankegel I believe the glibc package you need here is actually `libc6-dev-ppc64el-cross`

File: .github/workflows/pkgcheck.yml
Comment (mscastanho) [inline .github/workflows/pkgcheck.yml]: glibc packages on Ubuntu seem to start with `libc6`. So this should be `libc6-dev-powerpc-cross`

File: .github/workflows/pkgcheck.yml
Comment (mscastanho) [inline .github/workflows/pkgcheck.yml]: `libc6-dev-arm64-cross` here?

File: .github/workflows/pkgcheck.yml
Comment (mscastanho) [inline .github/workflows/pkgcheck.yml]: @dankegel Shouldn't you add `Ubuntu GCC PPC64LE` to this list too?

File: test/pkgcheck.sh
Comment (mscastanho) [inline test/pkgcheck.sh]: Should be `libc6-dev-ppc64el-cross`. And also replace `libc` by `libc6` on the other package names.

**Review by mscastanho — COMMENTED** [review]
@dankegel just some small comments about package names. I think the builds passed because the glibc cross packages are in the `recommends` list for the gcc package, but having the correct package names is ideal.
Also, your fix for the `CHOST` issue looks good. Now `uname` is correctly being detected as `linux` for those failing cases.

File: .github/workflows/pkgcheck.yml
Comment (dankegel) [reply .github/workflows/pkgcheck.yml]: Yes, looks like the libc- one is an alias, but it's always less confusing to use the real package name.

File: .github/workflows/pkgcheck.yml
Comment (dankegel) [reply .github/workflows/pkgcheck.yml]: thanks.  yesterday was not my best day :-)

File: .github/workflows/pkgcheck.yml
Comment (dankegel) [reply .github/workflows/pkgcheck.yml]: I used same package name as in cmake.yml.  Evidently the one missing the 6 is an alias, and it pulls in the right package.  Both files probably should use the real package name.

File: .github/workflows/pkgcheck.yml
Comment (dankegel) [reply .github/workflows/pkgcheck.yml]: yep.

**Comment by dankegel** [comment]
cherry-picking #668 and #669 still leaves us with configure applying -march=armv8-a+crc hither and yon for
arm-linux-gnueabihf, e.g. on adler32.c.

**Comment by nmoinvaz** [comment]
@dankegel are there also problems with configure using powerpc? I noticed we don't have a CI test for powerpc for configure..

**Comment by dankegel** [comment]
powerpc seems ok.  arm is more attached to -march flags than powerpc is, I think...

**Comment by dankegel** [comment]
The cmake O2/O3 issue seems beyond the goals of this pull request; my strategy of letting the wookie, er cmake, win seems like enough for the moment.

And I'm not sure this pull request should be blocked waiting for the 32 bit -march disagreement between cmake and configure to be solved...?  A later pull request can expand the list of architectures tested by pkgcheck.

Other than those, I think the pull request now covers the feedback received, and it's ready for merging.

File: configure
Comment (mtl1979) [inline configure:1183]: Short-circuiting the test should be reverted...

File: configure
Comment (mtl1979) [inline configure]: Assuming SSE4.2 is available is still bad as a lot of AMD processors don't have SSE4.2 as it was originally available on only Intel processors. I don't know when the test was removed as even my previous development machine didn't have SSE4.2 and zlib-ng worked on it using configure...

Comment (dankegel) [reply configure]: I think there was never a check in configure; I pulled in the check from CMakeLists.txt and
added it as a comment in configure.  I admit that's a bit confusing.  I suspect zlib-ng checks
for sse4.2 at runtime before using it...?

**Comment by nmoinvaz** [comment]
I think we should wait until some of the cmake/configure issues are resolved and then add more configurations to the pkgcheck. Sometimes these things take time. I have had PRs that take 3-4 months so I know how it is.

**Comment by dankegel** [comment]
h'ok.  Landing the other three pull requests first is fine.

**Comment by dankegel** [comment]
Let's do this :-)

**Comment by mtl1979** [comment]
This overlaps with #676 ?

**Comment by dankegel** [comment]
Only trivially.  git will do the right thing.  (I think.)

In retrospect, I suppose I should have waited until #676 was merged before saying "let's do this", to avoid that question.

(Merge could happen in either order, but if you merge this one first, #676 would be a no-op.)

**Comment by mtl1979** [comment]
As long as @Dead2 knows the correct order of merging pull requests...

**Comment by nmoinvaz** [comment]
@dankegel thank you for adding the new architectures. Are there any remaining outstanding issues that we haven't cataloged? I had also noticed the issue with `LDSHAREDLIBC` too during test.  And I also see that the line `LDFLAGS=${LDFLAGS-"-L."}` adds `-L.` which is not added in CMake.

File: config.sub
Comment (nmoinvaz) [inline config.sub]: Is it possible to do this inline in `configure` without a separate file?

File: config.sub
Comment (nmoinvaz) [inline config.sub]: I don't know that we need to mention GPL here.

File: test/pkgcheck.sh
Comment (nmoinvaz) [inline test/pkgcheck.sh]: Do we need this `SOURCE_DATE_EPOCH`? Maybe not..?

File: test/pkgcheck.sh
Comment (nmoinvaz) [inline test/pkgcheck.sh]: Is there something we can do about this? How many bytes were changed?

**Review by nmoinvaz — COMMENTED** [review]
I have given some feedback on a few things.

File: config.sub
Comment (dankegel) [reply config.sub]: Yes, but config.sub is traditional, and if somebody needs one of the arches not yet in this config.sub,
they can drop in the gnu one and get unblocked.  So there's utility in keeping it a separate file.

File: config.sub
Comment (dankegel) [reply config.sub]: Yeah, that could confuse license scanners.

File: test/pkgcheck.sh
Comment (dankegel) [reply test/pkgcheck.sh]: I believe I needed it on FreeBSD 12.1, Debian 8, and Ubuntu 14.04. 

See also https://reproducible-builds.org/docs/source-date-epoch/ for what this does.

File: test/pkgcheck.sh
Comment (dankegel) [reply test/pkgcheck.sh]: Per the dd command, 16 bytes.

One does not do anything about Apple tools, one endures them :-)

**Comment by dankegel** [comment]
-L. appears to do nothing and could be omitted.  I looked through git history and found it may have originally been
added in f6194ef3, but even then, it was a no-op (unless there is some strange ld out there
that searches the library search path even when you provide the filename of the library).

Likewise, the -L.. could be omitted from test/Makefile.in.

But as it does not cause configure and cmake to generate different bits,
I hesitate to remove it as part of this pull request... perhaps it could be a followon.

File: test/pkgcheck.sh
Comment (mtl1979) [reply test/pkgcheck.sh]: Would be nice to have some link to site that describes what those bytes actually are... Mach-O file format reference doesn't have clear definition what starts at byte offset 1337 as the file header isn't that long...

Comment (dankegel) [reply test/pkgcheck.sh]: Yes, that would be nice.  However, I don't know how to dig it up myself, and I don't know that it's worth holding up this pull request for.

**Comment by dankegel** [comment]
OK, I updated config.sub's comment to explain what it does better, and to
mention using autotools' config.sub as a workaround in case this one is too limited.

I also rebased.

I think that (and my replies above) resolves all the feedback received today sufficiently.
If the magic dd for macos seems unacceptable, I can remove macos support from pkgcheck.sh.

Please merge #676 first, then I'll rebase.

File: test/pkgcheck.sh
Comment (nmoinvaz) [reply test/pkgcheck.sh]: I'm not sure if this is current but `strip` seems to have a `-no_uuid` flag that will strip out all `LC_UUID` load commands.
https://www.manpagez.com/man/1/strip/

Comment (dankegel) [reply test/pkgcheck.sh]: That sounds handy.  Also, evidently ld has a -no_uuid option.  I'll see if those can replace the dd!

Comment (dankegel) [reply test/pkgcheck.sh]: So far no joy.  It was (and maybe is) a good theory, though.

I just refreshed the mac support a bit, turned up an interesting bit so I left it as a separate commit.

Comment (dankegel) [reply test/pkgcheck.sh]: Wait, there is joy in mudville.  Stand by...

Comment (dankegel) [reply test/pkgcheck.sh]: I got it to go green, but there's a catch.  See https://github.com/zlib-ng/zlib-ng/issues/681

File: test/pkgcheck.sh
Comment (nmoinvaz) [inline test/pkgcheck.sh]: We probably won’t need this after #674 is merged.

**Comment by dankegel** [comment]
h'ok, it passes on mac now, too, without the dastardly dd hack.

How do people feel about the way the mac AVX2 problem was
solved by overloading WITH_NATIVE_INSTRUCTIONS ?
I could have introduced a new option, but WITH_NATIVE_INSTRUCTIONS
is such a close match.

**Comment by dankegel** [comment]
Is that code coverage decrease real?

**Comment by Dead2** [comment]
This is now ready for rebase.

**Comment by mscastanho** [comment]
@dankegel I noticed that this leaves directories `btmp1`, `btmp2`, `pkgtmp1` and `pkgtmp2` after running the script. Maybe add them to `make clean`? Another option would be to erase them after the check is done, but it might be useful to be able to check them in case of error. This could easily be part of a future clean up PR as well.

**Comment by dankegel** [comment]
@mscastanho I added clean and .gitignore entries for pkgcheck's detritus, and made pkgcheck clean up after itself except on failure.

**Comment by dankegel** [comment]
Rebased and ready for final review and landing.

**Comment by dankegel** [comment]
The coverage changes seem spurious, but they're caused by something, and this pull request probably has to wait until it's understood.  It's mildly annoying that they only show up in the CI of the pull request, not the CI of the branch before creating the pull request...

**Comment by dankegel** [comment]
Let's land #692 first, then I'll rebase and figure out that coverage problem.

File: config.sub
Comment (nmoinvaz) [reply config.sub]: Is it possible to merge this file into the configure script?

Comment (nmoinvaz) [reply config.sub:1]: Is it possible to merge this file into the configure script?

File: configure
Comment (nmoinvaz) [inline configure:1124]: I assume with these changes now we always do SSE2 check for x64 instead of assuming it exists (like CMake) and that you have merged the x86 and x64 sections in configure?

**Review by nmoinvaz — COMMENTED** [review]
I have made only two comments. It looks much smaller indeed!

**Comment by nmoinvaz** [comment]
It seems I already asked the question about config.sub. I think it would be better if it was merged into configure in some way. That way configure script is more self contained. 

An alternative might be to create a “configure” directory same as the “CMake” directory and put it there but I’m not that fond of that idea either.

File: configure
Comment (dankegel) [reply configure:1124]: Yup.

File: test/pkgcheck.sh
Comment (dankegel) [reply test/pkgcheck.sh]: Gone.

**Comment by dankegel** [comment]
I removed the questionable set of CFLAGS in test/pkgcheck.sh, and moved config.sub out of the top level directory.

I think it really is good to go now...?

**Review by nmoinvaz — APPROVED** [review]
I think this commit makes the code better - I like how it no longer checks the CMakeCache.txt. Having the file in the _tools_ directory is a nicer too. @Dead2 will be able to speak to if he wants it there or in a _configure_ directory. Otherwise, LGTM

**Comment by dankegel** [comment]
Can't have it in a 'configure' subdirectory, there's a bit of a name clash there :-)

**Comment by dankegel** [comment]
Push the button @Dead2 :-)

**Comment by Dead2** [comment]
I am in no condition to review stuff right now, with 5 holes in my abdomen and recovering from surgery that did not go quite as planned (it turned out OK, but means a much longer and more painful recovery). I hope to be able to read without my eyes constantly crossing in a couple of days, after that I can probably get some more testing and reviewing done.

That said, this PR has been reviewed multiple times by me and others, and although I see only 1 current review approval, I'll merge this now so we can keep things moving along and get the rest rebased on top of this as well. The other PRs will have to wait a bit longer though, unless there are more-or-less trivial changes that get at least TWO review approvals in addition to me.

**Comment by dankegel** [comment]
Thank you and get well soon!

**Pull request merged**
